### ISSUE TITLE
[BUG] Epochs_trained gives out 0

### ISSUE THREAD

**Issue by yigitcancomlek** [unit8co/darts#1922]
**Describe the bug**
I am trying to export the number of epochs I have used during training my model before it stopped for validation loss. `model.epochs_trained` gives 0, at least for the RNNModel

**Expected behavior**
`model.epochs_trained` should give an integer with final number of epochs.

**System (please complete the following information):**
 - Python version: 3.8
 - darts version: 0.24.0

**Comment by dennisbader** [comment]
Hi @yigitcancomlek, I tried to reproduce your issue but for me everything works fine with the given setup.

```
>>> from darts.models import TFTModel
>>> from darts.datasets import AirPassengersDataset
>>> 
>>> s = AirPassengersDataset().load()
>>> m = RNNModel(input_chunk_length=12)
>>> m.fit(s, epochs=10)
>>> print(m.epochs_trained)
10
```

**Comment by yigitcancomlek** [comment]
Hello @dennisbader, thank you very much for your response! I assumed you had RNNModel imported from somewhere else as you initially imported TFTModel but used RNNmodel in the code.

Regarding the issue, here is a snippet of my results with your example

```
>>> from darts.models import TFTModel
>>> from darts.datasets import AirPassengersDataset
>>> 
>>> s = AirPassengersDataset().load()
>>> m = RNNModel(input_chunk_length=12)
>>> m.fit(s, epochs=10)
>>> print(m.epochs_trained)
10
>>> m.predict(series = s, n=1)
>>> print(m.epochs_trained)
0
```

Although this is not my exact application, I hope it still provides some clarification. For my own case, calling m.epochs_trained gives 0 right after model fitting, not sure why. Maybe I have val_series and EarlyStopping included in the fit, so when the model stops fitting early, it does not record the number of epochs?

Thanks a lot!

**Comment by dennisbader** [comment]
When you say right after training, do you mean right after prediction? In your example the epochs trained after training show `10`, only after prediction, they show `0`.
When predicting, a new PyTorch Lightning Trainer is created that runs the prediction. This is why the epochs trained fall back to 0 (the epochs are retrieved from the Trainer, so we get 0 because we create a new Trainer).

So you can only get the epochs right after training before running prediction.
It is the same when using EarlyStopping:

```
>>> from darts.models import RNNModel
>>> from darts.datasets import AirPassengersDataset
>>> from pytorch_lightning.callbacks import EarlyStopping
>>> 
>>> 
>>> es = EarlyStopping(monitor='val_loss', patience=3, min_delta=1e6)
>>> 
>>> s = AirPassengersDataset().load()
>>> m = RNNModel(
>>>     input_chunk_length=12,
>>>     pl_trainer_kwargs={"callbacks": [es]},
>>>     random_state=42,
>>> )
>>> m.fit(s, epochs=100, val_series=s[-25:])
>>> print(m.epochs_trained)
4
>>> m.predict(n=1, series=s)
>>> print(m.epochs_trained)
0
```


If you really need the epochs still after predictions, you could write a custom callback that counts the epochs:

```
>>> from darts.models import RNNModel
>>> from darts.datasets import AirPassengersDataset
>>> from pytorch_lightning.callbacks import EarlyStopping, Callback
>>> 
>>> 
>>> class MyEpochCounter(Callback):
>>>     def __init__(self):
>>>         self.epochs_trained = 0
>>> 
>>>     def on_train_epoch_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
>>>         self.epochs_trained += 1
>>> 
>>> 
>>> es = EarlyStopping(monitor='val_loss', patience=3, min_delta=1e6)
>>> my_counter = MyEpochCounter()
>>> 
>>> s = AirPassengersDataset().load()
>>> m = RNNModel(
>>>     input_chunk_length=12,
>>>     pl_trainer_kwargs={"callbacks": [es, my_counter]},
>>>     random_state=42,
>>> )
>>> m.fit(s, epochs=100, val_series=s[-25:])
>>> print(m.epochs_trained, my_counter.epochs_trained)
4, 4
>>> m.predict(n=1, series=s)
>>> print(m.epochs_trained, my_counter.epochs_trained)
0, 4
```

**Comment by yigitcancomlek** [comment]
Hello @dennisbader thank you very much for your response, this is super helpful. I should have included my code snippet here for more clarification. Since the backtesting function in darts does not seem to support hyperparameter tuning using an expanded window strategy, I tried to write my own loop that goes through the training data by expanding the training size by N points each iteration. 

In this optuna objective function loop shown below, I try to build and fit a model for each training pass while recording the errors on the validation set. Specifically, In the ` build_model` and `fit_model` function (defined below), I build and fit a new model using the new data. As you can see in the `fit_model` function, every time I call the  `print(model.epochs_trained)`, it results in 0. After your comments I wondered if I am overwriting  PyTorch Lightning Trainer in each iteration but even if that was the case, I should still get at least some number in the very first training instance. Is there a way to reset the models each time they are run in a loop? I thought the `force_reset` provides that capability for us.

```
def objective(trial):
    for k in range(pass_size):

        ## 
       time series and covariates are obtained, scaled here
       ##

        model_built = build_model(
        hyperparams
        )
        # fit the model here
        model_BT = fit_model(model = model_built,
                             ts_train=pass_train_y_scaled,
                             ts_val = pass_validation_y_scaled,
                             input_train=pass_train_x_scaled,
                             input_val=pass_validation_x_scaled,
                             )

        # make predictions here
        predictions = model_BT.predict(n=len(pass_validation_y_scaled),
                series=pass_train_y_scaled,
                future_covariates = pass_train_x_scaled.concatenate(pass_validation_x_scaled),
                num_samples = 1000
            )
```
Here is the fit_model function
```
def fit_model(model,ts_train,ts_val,input_train,input_val):    
    model.fit(
        series = ts_train,
        val_series = ts_val,
        future_covariates = input_train, # requires past inputs
        val_future_covariates=input_val # both past and future(validation inputs)
    )
    print(model.epochs_trained)
    # Reload best model 
    if ML_model == 'DeepAR':
        model =  RNNModel.load_from_checkpoint(study_name)
    return model
```

Here is the build_model function

```
def build_model(
    hyperparams,
    callbacks = None,
    tb_log = False,
):

    #Fixed parameters that will be the same for all models
    MAX_N_EPOCHS = 200

    # Monitor the validation loss for early stopping
    early_stopper = EarlyStopping("val_loss", patience=5, verbose=True) #check_on_train_epoch_end =True

    if callbacks is None:
        callbacks = [early_stopper]
    else:
        callbacks = [early_stopper] + callbacks

    pl_trainer_kwargs = {"callbacks": callbacks}

    # building the ML models here 
    if ML_model == 'DeepAR':
        model_built = RNNModel(
            model='LSTM',
            training_length = val_length-1,
            input_chunk_length=hyperparams['input_chunk_length'],
            n_rnn_layers=hyperparams['n_rnn_layers'],
            hidden_dim=hyperparams['hidden_dim'],
            batch_size = hyperparams['batch_size'],
            n_epochs = MAX_N_EPOCHS,
            dropout=hyperparams['dropout'],
            optimizer_kwargs={"lr":1e-4},
            optimizer_cls = torch.optim.Adam,
            lr_scheduler_kwargs= {"optimizer": torch.optim.Adam},
            lr_scheduler_cls = torch.optim.lr_scheduler.ReduceLROnPlateau,
            likelihood=GaussianLikelihood(),
            pl_trainer_kwargs=pl_trainer_kwargs,
            model_name=study_name,
            force_reset=True,
            save_checkpoints = True,
            log_tensorboard = tb_log,
            torch_metrics = torch_metrics,
            random_state = 42
        )
    return model_built
```
